tests/governance_js.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import infra.network
+import infra.path
+import infra.proc
+import infra.net
+import infra.e2e_args
+import suite.test_requirements as reqs
+
+
+def action(name, **args):
+    return {"name": name, "args": args}
+
+
+@reqs.description("Test proposal validation")
+def test_proposal_validation(network, args):
+    primary, _ = network.find_nodes()
+    add_member = [action("add_member", cert="", enc_pubk="", member_data={})]
+
+    valid_set_recovery_threshold = [action("set_recovery_threshold", threshold=5)]
+    no_args_set_recovery_threshold = [action("set_recovery_threshold")]
+    bad_arg_set_recovery_threshold = [action("set_recovery_threshold", threshold=5000)]
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", valid_set_recovery_threshold)
+        assert r.status_code == 200, r.body.text()
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", valid_set_recovery_threshold * 2)
+        assert r.status_code == 200, r.body.text()
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", no_args_set_recovery_threshold)
+        assert r.status_code == 400 and r.body.json()["error"]["code"] == "ProposalFailedToValidate", r.body.text()
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", no_args_set_recovery_threshold + bad_arg_set_recovery_threshold)
+        assert r.status_code == 400 and r.body.json()["error"]["code"] == "ProposalFailedToValidate", r.body.text()
+
+    return network
+
+@reqs.description("Test proposal storage")
+def test_proposal_storage(network, args):
+    primary, _ = network.find_nodes()
+    add_member = [action("add_member", cert="", enc_pubk="", member_data={})]
```

Comment:
Should probably be wrapped in `{"actions": ...}` for future extensibility.

Reply:
Yes, good point.